library/core/src/iter/adapters/array_chunks.rs

```diff
@@ -0,0 +1,182 @@
+            match self.iter.next_chunk() {
+                Ok(chunk) => acc = f(acc, chunk)?,
+                Err(remainder) => {
+                    // Make sure to not override `self.remainder` with an empty array
+                    // when `next` is called after `ArrayChunks` exhaustion.
+                    self.remainder.get_or_insert(remainder);
+
+                    break try { acc };
+                }
+            }
+        }
+    }
+
+    fn fold<B, F>(mut self, init: B, f: F) -> B
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> B,
+    {
+        self.try_fold(init, NeverShortCircuit::wrap_mut_2(f)).0
+    }
+}
+
+#[unstable(feature = "iter_array_chunks", reason = "recently added", issue = "100450")]
+impl<I, const N: usize> DoubleEndedIterator for ArrayChunks<I, N>
+where
+    I: DoubleEndedIterator + ExactSizeIterator,
+{
+    #[inline]
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.try_rfold((), |(), x| ControlFlow::Break(x)).break_value()
+    }
+
+    fn try_rfold<B, F, R>(&mut self, init: B, mut f: F) -> R
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> R,
+        R: Try<Output = B>,
+    {
+        // We are iterating from the back we need to first handle the remainder.
+        self.next_back_remainder();
+
+        let mut acc = init;
+        let mut iter = ByRefSized(&mut self.iter).rev();
+
+        // NB remainder is handled by `next_back_remainder`, so
+        // `next_chunk` can't return `Err` with non-empty remainder
+        // (assuming correct `I as ExactSizeIterator` impl).
+        while let Ok(mut chunk) = iter.next_chunk() {
+            // FIXME: do not do double reverse
+            //        (we could instead add `next_chunk_back` for example)
+            chunk.reverse();
+            acc = f(acc, chunk)?
+        }
+
+        try { acc }
+    }
+
+    fn rfold<B, F>(mut self, init: B, f: F) -> B
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> B,
+    {
+        self.try_rfold(init, NeverShortCircuit::wrap_mut_2(f)).0
+    }
+}
+
+impl<I, const N: usize> ArrayChunks<I, N>
+where
+    I: DoubleEndedIterator + ExactSizeIterator,
+{
+    /// Updates `self.remainder` such that `self.iter.len` is divisible by `N`.
+    fn next_back_remainder(&mut self) {
+        // Make sure to not override `self.remainder` with an empty array
+        // when `next_back` is called after `ArrayChunks` exhaustion.
+        if self.remainder.is_some() {
+            return;
+        }
+
+        // We use the `ExactSizeIterator` implementation of the underlying
+        // iterator to know how many remaining elements there are.
+        let rem = self.iter.len() % N;
+
+        // Take the last `rem` elements out of `self.iter`.
+        let mut remainder =
+            // SAFETY: `unwrap_err` always succeeds because x % N < N for all x.
+            unsafe { self.iter.by_ref().rev().take(rem).next_chunk().unwrap_err_unchecked() };
```

Comment:
I don't think I'll block the PR on this, but `unwrap_err_unchecked` here scares me -- I added a note to the tracking issue.

My instinct for now is that `%N` is definitely trustworthy, so that combined with `take` is sufficient to say this is ok.  But I'm still afraid, since it's not `TrustedLen`, and wonder if someone can come up with an evil example here where it somehow gets a full chunk and thus is UB in safe code.

(At least the obvious things, like a `len()` that always says `usize`, is protected against because of the modulo.)

Reply:
As already noted in another comment I don't like this entire method anyway. Either we should add `next_chunk_back` or remove the DEI impl.

Reply:
So, an evil impl can return less than `rem` elements, but `actual <= rem < N` still holds so the chunk will be an error no matter what.

An evil impl can just return a wrong length, so after this iterator returns a number of elements not divisible by `N`. But this is fine too, an `Err` will just be ignored:
https://github.com/rust-lang/rust/blob/5fbcde1b55c09421f43a7d6cfe09103d064ed2db/library/core/src/iter/adapters/array_chunks.rs#L116-L119

Reply:
That being said, as @the8472 already said, we should probably rewrite the DEI impl anyway (I'll try to do this, after this lands).

Reply:
Yeah, the `%` is certainly fine.  What I'm not yet convinced of is that there's no way to make `take` misbehave somehow -- for example, `Take::try_fold` uses the inner `try_fold`, so maybe there'd be a way for that to be implemented wrongly-but-not-UB that could result in too many things getting put in the array somehow.

But I agree that just making a nicer implementation without the `unsafe{}` is the right plan.

Reply:
Actually you are right 😨 

Overriding `try_fold` with some nasty unsafe to skip over `Break(_)` allows you to trick `Take` into returning more elements that it should ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=223410f4b12798cd5490928290de878f)).

`n` can underflow, if a bad iterator impl skips over `Break(_)`:
https://github.com/rust-lang/rust/blob/2fbc08e2ce64dee45a29cb6133da6b32366268aa/library/core/src/iter/adapters/take.rs#L87-L89

The fact that unsafe code can't trust `Take` is scary.

Reply:
I think the problem there is still in the `unsafe` in `Misbehave` -- it protects against double-drop, but duplicating arbitrary values can violate *safety* invariants in general.  (For example, if the type is `!Clone + !Default`, I can use you moving it into me to track resource consumption, like a ZST tracker for a global semaphore.)  So that specific example is just "well unsound `unsafe`-using code breaks everything".

But yeah, even though I've not been able to come up with a safe trick that would do it, things along those lines are what make me worried about it.  (And certainly if specialization existed then it would be possible to do particularly weird things in safe code because it can violate parametricity <sub>even worse than in normal code</sub>.)

Reply:
The example could check, that `type_id::<B>() == type_id::<usize>()` or something similar, which would IMO make the code sound.

Reply:
Yeah, I thought about that one, but `TypeId::of::<B>()` needs `B: 'static`, which we don't have here.  Regardless, it's yet another chink in the armour that makes be scared we'll break through eventually.